engine/consensus/compliance/core.go

```diff
@@ -280,26 +315,49 @@ func (c *Core) processBlockAndDescendants(proposal *messages.BlockProposal) erro
 // Expected errors during normal operations:
 //   - engine.OutdatedInputError if the block proposal is outdated (e.g. orphaned)
 //   - engine.InvalidInputError if the block proposal is invalid
-func (c *Core) processBlockProposal(proposal *messages.BlockProposal) error {
+//   - engine.UnverifiableInputError if the proposal cannot be validated
+func (c *Core) processBlockProposal(proposal *messages.BlockProposal, parent *flow.Header) error {
 	startTime := time.Now()
 	defer c.complianceMetrics.BlockProposalDuration(time.Since(startTime))
 
 	header := proposal.Header
-	id := header.ID()
+	blockID := header.ID()
 
-	span, ctx, isSampled := c.tracer.StartBlockSpan(context.Background(), id, trace.ConCompProcessBlockProposal)
+	span, ctx, isSampled := c.tracer.StartBlockSpan(context.Background(), blockID, trace.ConCompProcessBlockProposal)
 	if isSampled {
 		span.SetAttributes(
 			attribute.String("proposer", header.ProposerID.String()),
 		)
 	}
 	defer span.End()
 
+	hotstuffProposal := model.ProposalFromFlow(header, parent.View)
+	err := c.validator.ValidateProposal(hotstuffProposal)
```

Comment:
The processBlockProposal will not be called concurrently right? because the engine's `processAvailableMessages` method delivers one message at a time, right?

Since `processBlockProposal` is called in a single thread manner, then this `ValidateProposal` is also called in a single thread manner right?

Before the validation is done in hotstuff in a different thread, then the payload validation (state.Extend) and header validation (validator.ValidateProposal) are done concurrently. 

Note what's concurrently validated is not for the same block, but different block, meaning it's concurrently validating a payload for one block, and validating a header for a different block. 

Now in this PR, it seems both the payload and header validation are done by compliance sequentially, as there is no concurrent validation. Wouldn't it slow down the process of a proposal?

Reply:
Yes, excellent point. There is a subtle concurrency we're losing here. The reason we could extract that concurrency previously is that we stored potentially invalid blocks, and validated payloads before signatures, which we would like to avoid doing. 

Earlier in this feature branch, we added the ability to [query identities by view, for known epochs](https://github.com/onflow/flow-go/blob/6a39c8c64184231a05dd5c756db91307f0361e78/consensus/hotstuff/committee.go#L81). 

With the change to front-load QC verification, conceptually, we can perform different concurrent processing:
1. We can validate block payloads (compliance checks) in different forks concurrently, so long as we validate parents before children. 
  * This requires ensuring `MutableState` is concurrency safe
  * This only helps in the unhappy path, when we have competing forks
2. We can validate block headers (HotStuff checks) all concurrently, so long as they refer to a view for a known epoch (true the vast majority of the time). 
  * To do this, we need to know the parent view of the block, which we currently only discover by directly observing and processing the parent block itself (ie. in practice we need to process the parent block first, anyway, so we lose most of the concurrency possibilities)

Reply:
In my mind we would to take a gradual approach for this problem. First we update compliance engine to correctly process blocks including header validation, payload validation, state extension and passing to hotstuff. We do this as Jordan did this. In single threaded fashion. Once we verify our implementation is correct and actually works we take advantage of stateless header validation and build a pipelines where we validate headers concurrently and then pass them to single-threaded pipeline where we validate payload and apply state changes in strict parent-child order. 
This approach doesn't get a speedup when producing blocks in steady pace but gives speedup when blocks are delivered faster than we can process them(in case of sync for instance). 
After this we can think about concurrent payload validation but it gets tricky at this point and requires careful organization due to parent-child requirement. Might not worth the effort.

To summarize, I am happy how Jordan approached this, we should incrementally update our implementation. In next iterations we will be adding concurrency support at least for header validation.

Reply:
Mentioning the following just for completeness. Conceptually, there is an even more efficient approach possible:
* Concurrency mainly pays off, if a node is catching up and has a longer backlog of blocks pending validation
* In many cases, we will receive such blocks from the sync engine in batches, such as `A <- B <- C <- D <- E`
* Technically, it is sufficient to validate
   - validate the hashes of the blocks and payload for all blocks `A, .. , E`
   - validate the QC from block `E` (pointing to block `D`)
   - validate payload of block `E`
   
   This is sufficient because consensus is an inductive protocol. Since block `D` received a QC, we  can be certain that all ancestor blocks up to and including `D` must be correct, including their payload. Hence, we could just write blocks `A, .. , D` into storage without _any_ validation (except for recomputing the hashes to confirm integrity of the blocks)

Note:
* In this approach, we firmly assume that byzantine nodes will never reach a supermajority of stake (_more_ than 2/3). If this assumption is violated, an honest node that is catching up might not notice and erroneously vote for an extension of an invalid fork. 
* Nevertheless, safety of finality already breaks at byzantine nodes reaching 1/3 of stake, at which double-spend attacks become theoretically possible. 

However, _exceeding_ 2/3 of Byzantine stake is a much larger hurdle for an attacker. Hence, I feel the assumption of having at most 2/3 byzantine stake is fine because other important consensus guarantees break much earlier.

Reply:
I think **practically most impactful** would be concurrent validation of the _same block_. Reasoning:
* Concurrent header and payload validation of _different_ blocks is mainly useful for nodes that are catching up. I consider it unlikely that this parallelism will result in noteworthy performance improvements on the happy path. 
* In comparison, validating different aspects of the _same_ proposal concurrently has the potential to dramatically speed up the consensus hotpath and thereby improve block rate and reduce platform latency overall. A block has many different aspects that could be checked in parall:
  * header, incl. proposer signature, height, view etc
  * QC for parent
  * seals (validity of seals only depends on ancestor blocks but not on the current block's payload)
  * execution results & receipts 
  * collection guarantees

  At the moment, we check these 5 aspects sequentially. I assume some will be much more expensive to check, but if we check all 5 aspects in parallel, we could potentially still cut verification time down to half.

  With similar improvements for block production, we could reach 1s block time (with a multitude of further improvements also being possible).